internal/webhook/external/github.go

```diff
@@ -6,57 +6,66 @@ import (
 	"net/http"
 
 	gh "github.com/google/go-github/v71/github"
-	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/git"
 	xhttp "github.com/akuity/kargo/internal/http"
 	"github.com/akuity/kargo/internal/io"
 	"github.com/akuity/kargo/internal/logging"
 )
 
-// githubHandler handles push events for github.
-// After the request has been authenticated,
-// the kubeclient is queried for all warehouses that contain a subscription
-// to the repo in question. Those warehouses are then patched with a special
-// annotation that signals down stream logic to refresh the warehouse.
-func githubHandler(
+// githubWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from GitHub.
+type githubWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newGitHubWebhookReceiver returns a new instance of githubWebhookReceiver.
+func newGitHubWebhookReceiver(
 	c client.Client,
-	namespace string,
-	secretName string,
-) http.HandlerFunc {
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &githubWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.GitHub.SecretRef.Name,
+		},
+	}
+}
+
+// GetDetails implements WebhookReceiver.
+func (g *githubWebhookReceiver) getReceiverType() string {
+	return "github"
+}
+
+// getSecretValues implements WebhookReceiver.
+func (g *githubWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	token, ok := secretData["token"]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for a GitHub WebhookReceiver")
+	}
+	return []string{string(token)}, nil
+}
+
+// GetHandler implements WebhookReceiver.
+func (g *githubWebhookReceiver) GetHandler() http.HandlerFunc {
```

Comment:
It feels like this could be broken up even more, e.g.

```go
func (g *githubWebhookReceiver) GetHandler() http.HandlerFunc {
    return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
        ctx := r.Context()
        logger := logging.LoggerFromContext(ctx)

        if err := g.validateRequest(r); err != nil {
            g.handleError(w, err)
            return
        }

        event, err := g.parseWebhook(r)
        if err != nil {
            g.handleError(w, err)
            return
        }

        g.handleEvent(ctx, w, event)
    })
}
```